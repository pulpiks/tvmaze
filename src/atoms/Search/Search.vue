<template>
  <b-form ref="form" @submit="onSubmitSearch" @reset="onReset">
    <b-container
      fluid
      no-gutters
      :class="{
        searchContainer: true,
        'py-0': true,
        'px-0': true,
        'bg-white': true
      }"
    >
      <b-row class="d-flex w-100" no-gutters>
        <b-col :cols="isSubmitBtnVisible ? 10 : 12">
          <div class="search__wrap d-flex align-items-center border rounded p-1 bg-light">
            <b-icon icon="search" variant="dark" font-scale="1" class="d-flex mx-2"></b-icon>
            <b-form-input
              :value.sync="search"
              type="text"
              class="searchInput pl-2 pr-2 pb-0 pt-0"
              :placeholder="placeholder"
              data-test="search-input"
              @input="onChangeValue"
            ></b-form-input>
            <button type="button" class="search__btn-close ml-2"
              aria-label="Close" :disabled="isCloseBtnDisabled"
                @click="clearSearch" data-test="search-btn-close">
              <span aria-hidden="true">&times;</span>
            </button>
          </div>
        </b-col>
        <b-col class="ml-1" v-if="isSubmitBtnVisible" data-test="search-btn-submit">
          <b-button
            v-if="isSubmitBtnVisible"
            :disabled="isSubmitBtnDisabled"
            variant="primary"
            class="w-100"
            type="submit"
            >Search</b-button
          >
        </b-col>
      </b-row>
    </b-container>
  </b-form>
</template>
<script lang="ts">
import { Component, Vue, Prop } from 'vue-property-decorator'

export type VForm = Vue & {
  validate: () => boolean
  resetValidation: () => boolean
  reset: () => void
 }

@Component({
  name: 'Search'
})

export default class extends Vue {
  @Prop({ required: false }) private fullWidth!: boolean

  @Prop({ required: false }) private className!: string

  @Prop({ required: false }) private placeholder!: string

  @Prop({ required: true }) private value!: string

  @Prop({ required: false }) private onChange!: (values?: any) => void;

  @Prop({ required: false }) private onSubmit!: (values?: any) => void;

  get search() {
    return this.value
  }

  get isSubmitBtnVisible() {
    return typeof this.$listeners.onSubmit === 'function'
  }

  get isSubmitBtnDisabled() {
    return this.isSubmitBtnVisible && !this.search
  }

  get isCloseBtnDisabled() {
    return !this.search
  }

  onSubmitSearch(evt: Event) {
    evt.preventDefault()
    if (this.$listeners.onSubmit) {
      this.$emit('onSubmit', this.search)
    }
  }

  onReset() {
    this.$emit('onReset', this.search)
  }

  onChangeValue(v: string) {
    if (this.$listeners.onChange) {
      this.$emit('onChange', v)
    }
  }

  clearSearch() {
    (this.$refs.form as VForm).reset()
  }
}
</script>
<style scoped>


.searchInputWrap {
  position: relative;
}
.searchInput {
  font-size: 20px;
  color: #222222;
  border: 0;
  height: 60px;
}

.searchInputWrap .searchIcon {
  left: 0;
}
.search__btn-close {
  right: 0;
  cursor: pointer;
}
</style>
