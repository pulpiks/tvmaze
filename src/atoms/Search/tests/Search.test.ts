import VueRouter from 'vue-router'
import { BootstrapVue, BootstrapVueIcons, BForm } from 'bootstrap-vue'
import {
  createLocalVue, shallowMount, VueClass, Wrapper, mount
} from '@vue/test-utils'

const localVue = createLocalVue()
localVue.use(VueRouter)
localVue.use(BootstrapVue)
localVue.use(BootstrapVueIcons)

// eslint-disable-next-line import/order, import/first
// eslint-disable-next-line import/first
import Search from '../Search.vue'

const testAnchors = {
  inputText: 'search-input',
  btnClose: 'search-btn-close',
  btnSubmit: 'search-btn-submit'
}


const setUp = (props = {}) => {
  const wrapper = shallowMount(Search, {
    stubs: ['b-icon', 'b-form-input', 'b-col', 'b-form', 'b-button', 'b-container', 'b-row', 'b-img'],
    propsData: props
  })
  return wrapper
}


describe('Search component', () => {
  let wrapper: Wrapper<Search>

  beforeAll(() => {
    localVue.component('b-form', BForm)
  })

  afterAll(() => {
    wrapper.destroy()
  })

  const props = {
    placeholder: 'Search by name',
    value: 'text',
    onChange: jest.fn(),
    onReset: jest.fn()
  }

  it('render snapshot correctly', () => {
    wrapper = setUp(props)
    expect(wrapper.element).toMatchSnapshot()
  })

  it('render all basic show`s attributes correctly', () => {
    wrapper = setUp(props)
    const textInput = wrapper.find(`[data-test='${testAnchors.inputText}']`)
    const btnClose = wrapper.find(`[data-test='${testAnchors.btnClose}']`)
    const btnSubmit = wrapper.find(`[data-test='${testAnchors.btnSubmit}']`)
    expect(textInput.exists()).toBe(true)
    expect(btnClose.exists()).toBe(true)
    expect(btnSubmit.exists()).toBe(false)
  })

  it('callback is invoked when value is changed', async () => {
    wrapper = mount(Search, {
      propsData: props,
      localVue,
      listeners: {
        onChange: props.onChange
      }
    })
    const inputEl = wrapper.find('input[type="text"]')
    inputEl.setValue('itworks')
    await inputEl.trigger('input')
    expect((inputEl.element as any).value).toContain('itworks')
    expect(props.onChange).toBeCalled()
  })
})
