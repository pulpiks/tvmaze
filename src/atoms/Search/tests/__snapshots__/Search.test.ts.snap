// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component render snapshot correctly 1`] = `
<b-form-stub onreset="function() {
        return fn.apply(this, arguments);
      }">
  <b-container-stub fluid="" no-gutters="" class="searchContainer py-0 px-0 bg-white">
    <b-row-stub no-gutters="" class="d-flex w-100">
      <b-col-stub cols="12">
        <div class="search__wrap d-flex align-items-center border rounded p-1 bg-light">
          <b-icon-stub icon="search" variant="dark" font-scale="1" class="d-flex mx-2"></b-icon-stub>
          <b-form-input-stub value="text" type="text" placeholder="Search by name" data-test="search-input" class="searchInput pl-2 pr-2 pb-0 pt-0"></b-form-input-stub> <button type="button" aria-label="Close" data-test="search-btn-close" class="search__btn-close ml-2"><span aria-hidden="true">Ã—</span></button>
        </div>
      </b-col-stub>
      <!---->
    </b-row-stub>
  </b-container-stub>
</b-form-stub>
`;
